pkg/crosscluster/logical/table_batch_handler_test.go

```diff
@@ -0,0 +1,405 @@
+
+	eb := newEventBuilder(t, runner, s)
+
+	handler := newTestTableHandler(t, eb.descID, s)
+
+	type previousValue string
+	const (
+		previousValuePresent previousValue = "present"
+		previousValueAbsent  previousValue = "absent"
+	)
+	previousValues := []previousValue{
+		previousValuePresent,
+		previousValueAbsent,
+	}
+
+	type localValue string
+	const (
+		// The local value is the same as the previous value supplied by the
+		// event.
+		localValueMatch localValue = "previous-value"
+		// The local value is different from the previous value supplied by the
+		// event.
+		localValueMismatch localValue = "different-value"
+		// The local value never existed in the database.
+		localValueNull localValue = "null-value"
+		// The local value is a tombstone.
+		localValueTombstone localValue = "tombstone-value"
+	)
+	localValues := []localValue{
+		localValueMatch,
+		localValueMismatch,
+		localValueNull,
+		localValueTombstone,
+	}
+
+	type replicationType string
+	const (
+		replicationTypeUpdate replicationType = "update"
+		replicationTypeInsert replicationType = "insert"
+		replicationTypeDelete replicationType = "delete"
+	)
+	replicationTypes := []replicationType{
+		replicationTypeUpdate,
+		replicationTypeInsert,
+		replicationTypeDelete,
+	}
+
+	start := s.Clock().Now()
+
+	type testCase struct {
+		id              int
+		previousValue   previousValue
+		localValue      localValue
+		replicationType replicationType
+		winLww          bool
+	}
+	testCaseName := func(tc testCase) string {
+		return fmt.Sprintf("replication-type=%s:previous-value=%s:local-value=%s:win-lww=%t", tc.replicationType, tc.previousValue, tc.localValue, tc.winLww)
+	}
+	getPreviousValue := func(tc testCase) string {
+		if tc.previousValue == previousValuePresent {
+			return fmt.Sprintf("previous-value-%s", testCaseName(tc))
+		}
+		return ""
+	}
+	getNewValue := func(tc testCase) string {
+		return fmt.Sprintf("new-value-%s", testCaseName(tc))
+	}
+	getDecodedEvent := func(tc testCase) decodedEvent {
+		origin := start
+		if tc.winLww {
+			origin = s.Clock().Now()
+		}
+		switch tc.replicationType {
+		case replicationTypeUpdate:
+			return eb.makeUpdate(fmt.Sprintf("%d", tc.id), getPreviousValue(tc), getNewValue(tc), origin)
+		case replicationTypeInsert:
+			return eb.makeInsert(fmt.Sprintf("%d", tc.id), getNewValue(tc), origin)
+		case replicationTypeDelete:
+			return eb.makeDelete(fmt.Sprintf("%d", tc.id), getPreviousValue(tc), origin)
+		default:
+			panic(fmt.Sprintf("unknown replication type: %s", tc.replicationType))
+		}
+	}
+	getLocalValue := func(tc testCase) string {
+		switch tc.localValue {
+		case localValueMatch:
+			return getPreviousValue(tc)
+		case localValueMismatch:
+			return fmt.Sprintf("value-mismatch-%d", tc.id)
+		}
+		return ""
+	}
+
+	// Construct the Cartesian product of all the possible test cases.
+	var testCases []testCase
+	id := 1337 // Start at a non-zero value to avoid bugs that are papered over by ordered ids.
```

Comment:
i realize this doesn't matter much, but i'm curious. what bugs are we trying to avoid here?